nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-client-service/src/main/java/org/apache/nifi/pulsar/StandardPulsarClientService.java

```diff
@@ -0,0 +1,285 @@
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("30 sec")
+            .build();
+
+    public static final PropertyDescriptor LISTENER_THREADS = new PropertyDescriptor.Builder()
+            .name("Listener Threads")
+            .description("The number of threads to be used for message listeners")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("1")
+            .build();
+
+    public static final PropertyDescriptor MAXIMUM_LOOKUP_REQUESTS = new PropertyDescriptor.Builder()
+            .name("Maximum lookup requests")
+            .description("Number of max lookup-requests allowed on each broker-connection. To prevent overload on broker,"
+                    + "it should be greater than the 'Maximum concurrent lookup-requests' property value.")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("50000")
+            .build();
+
+    public static final PropertyDescriptor MAXIMUM_REJECTED_REQUESTS = new PropertyDescriptor.Builder()
+            .name("Maximum rejected requests per connection")
+            .description("Max number of broker-rejected requests in a certain time-frame after " +
+                "which current connection will be closed and client creates a new connection that give " +
+                "chance to connect a different broker")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("50")
+            .build();
+
+    public static final PropertyDescriptor OPERATION_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("Operation Timeout")
+            .description("Producer-create, subscribe and unsubscribe operations will be retried until this " +
+                "interval, after which the operation will be marked as failed")
+            .required(false)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .defaultValue("30 sec")
+            .build();
+
+    public static final PropertyDescriptor STATS_INTERVAL = new PropertyDescriptor.Builder()
+            .name("Stats interval")
+            .description("The interval between each stat infomation update. It should be set to at least 1 second")
+            .required(false)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .defaultValue("60 sec")
+            .build();
+
+    public static final PropertyDescriptor TLS_TRUST_CERTS_FILE_PATH = new PropertyDescriptor.Builder()
```

Comment:
if the SSL Context Service is being passed in this should not be necessary.

Reply:
The SSL Context service only encapsulates two value, a KeyStore file and a TrustStore file, which I use to configure the broker endpoint. But there is also another property used to set the path to the trusted TLS certificate file (for the X509 certificate, etc) the broker uses, so I had to expose that setting in a separate property.

https://pulsar.incubator.apache.org/api/client/ has the detailed explanations for these properties.